letter-sender/src/main/java/com/christmas/letter/sender/service/SnsService.java

```diff
@@ -0,0 +1,53 @@
+package com.christmas.letter.sender.service;
+
+import com.christmas.letter.sender.exceptions.InvalidDataException;
+import com.christmas.letter.sender.model.Letter;
+
+import java.util.regex.Pattern;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import software.amazon.awssdk.services.sns.SnsClient;
+import software.amazon.awssdk.services.sns.model.PublishRequest;
+
+@Service
+@RequiredArgsConstructor
+public class SnsService {
+
+  @NonNull
+  private final SnsClient snsClient;
+  private final String snsTopicArn = "arn:aws:sns:us-east-1:000000000000:ChristmasLetterSns";
+  private static final Pattern EMAIL_PATTERN = Pattern.compile(
+      "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+\\.[a-zA-Z]{2,6}$");
+
+  public String publishLetter(Letter letter) {
+
+      if (letter.getEmail() == null || !EMAIL_PATTERN.matcher(letter.getEmail()).matches()) {
```

Comment:
Let's move the validations outside the service layer. You can specify the regex along with the nonnull validation in the model itself, via annotations.
See https://www.javaguides.net/2021/04/spring-boot-dto-validation-example.html

Reply:
We already have these validations on the Letter object, do we need to do the same check here as well?  
@NotNull(message="Email required")
  @Email(regexp = "[a-z0-9._%+-]+@[a-z0-9.-]+\\.[a-z]{2,3}")